frontend/src/InputForm.js

```diff
@@ -0,0 +1,546 @@
+              )
+              : section)})
+          : course)
+    }))
+  }
+
+  updateTimeStartPreference(id, timePreferenceStartTime) {
+    this.setState(state => ({
+      ...this.state,
+      criterion: {
+        ...state.criterion,
+        timePreferences: this.state.criterion.timePreferences.map((timePreference, index) =>
+          index === id
+            ? ({...timePreference, startTime: timePreferenceStartTime})
+            : timePreference)
+      }
+    }))
+  }
+
+  updateTimeEndPreference(id, timePreferenceEndTime) {
+    this.setState(state => ({
+      ...this.state,
+      criterion: {
+        ...state.criterion,
+        timePreferences: this.state.criterion.timePreferences.map((timePreference, index) =>
+          index === id
+            ? ({...timePreference, endTime: timePreferenceEndTime})
+            : timePreference)
+      }
+    }))
+  }
+
+  updateSubjectPreference(preferredSubject) {
+    this.setState(state => ({
+      ...this.state,
+      criterion: {
+        ...state.criterion,
+        preferredSubject: preferredSubject
+      }
+    }))
+  }
+  
+  updateMinCredits(minCredits) {
+    this.setState(state => ({
+      ...this.state,
+      basicInfo: {
+        ...state.basicInfo,
+        credits: {
+          ...state.basicInfo.credits, 
+          minCredits: minCredits
+        }
+      }
+    }))
+  }
+
+  updateMaxCredits(maxCredits) {
+    this.setState(state => ({
+      ...this.state,
+      basicInfo: {
+        ...state.basicInfo,
+        credits: {
+          ...state.basicInfo.credits, 
+          maxCredits: maxCredits
+        }
+      }
+    }))
+  }
+
+  updateTermStartDate(termStartDate) {
+    this.setState(state => ({
+      ...this.state,
+      basicInfo: {
+        ...state.basicInfo,
+        termDates: {
+          ...state.basicInfo.termDates, 
+          startDate: termStartDate
+        }
+      }
+    }))
+  }
+
+  updateTermEndDate(termEndDate) {
+    this.setState(state => ({
+      ...this.state,
+      basicInfo: {
+        ...state.basicInfo,
+        termDates: {
+          ...state.basicInfo.termDates, 
+          endDate: termEndDate
+        }
+      }
+    }))
+  }
+
+  createNewCourse() {
+    const defaultCourse = {
+      name: "course name",
+      courseID: "",
+      subject: "",
+      credits: 0,
+      isRequired: false,
+      rank: "",
+      sections: [{
+        professor: "",
+        startTime: "",
+        endTime: "",
+        days: []
+      }]
+    }
+
+    this.setState({
+      ...this.state,
+      courses: this.state.courses.concat(defaultCourse)
+    });
+  }
+
+  updateRankSelectOptions(id) {
+    let items = [];   
+    let numCourses = this.state.courses.length;     
+    if (this.state.courses[id].rank !== "") {
+      for (let j = 1; j <= numCourses; j++) {
+        items.push(<MenuItem key={j} value={j}>{j}</MenuItem>)       
+      }
+      return items;
```

Comment:
It works how I intend, but my intention might not be ideal, this was my thinking: If you haven't selected a rank, you can only choose from the available ranks (that haven't been chosen in other courses). If you have selected a rank though, any option is open in case you want to switch the order, and if you select a rank that's been chosen, the other course with that rank will be cleared. (If it weren't this way, once you pick a rank you wouldn't be able to change it). If there seems like a better option though, I'd love to hear it!

Reply:
I could make it so that there's a "none" option, or similar, for switching around the ranks if you think that'd be better.

Reply:
update: we're actually going to change the ranking logic so that it's a "score" from 1-5 of how much you prefer that course, so courses can have the same score